spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,89 +113,251 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
+			Assert.isTrue(StringUtils.hasText(output.value()), "At least one output must be specified");
+			return StringUtils.hasText(output.value()) ? output.value() : null;
+		}
+		return null;
+	}
+
+	private void assertStreamListenerMethod(int inputAnnotationCount, int outputAnnotationCount,
+			String methodAnnotatedInboundName, String methodAnnotatedOutboundName, boolean isDeclarative) {
+		if (!isDeclarative) {
+			Assert.isTrue(inputAnnotationCount == 0 && outputAnnotationCount == 0,
+					"@Input or @Output annotation is not supported as method parameter in StreamListener method with " +
+							"message handler mapping");
+		}
+		if (StringUtils.hasText(methodAnnotatedInboundName) && StringUtils.hasText(methodAnnotatedOutboundName)) {
+			Assert.isTrue(inputAnnotationCount == 0 && outputAnnotationCount == 0, "@Input or @Output annotations are not permitted as " +
+					"method parameters when both inbound and outbound values are set as method annotated values");
+		}
+		if (StringUtils.hasText(methodAnnotatedInboundName)) {
+			Assert.isTrue(inputAnnotationCount == 0, "Cannot set both StreamListener value " + methodAnnotatedInboundName +
+					" and @Input annotation as method parameter");
+			Assert.isTrue(outputAnnotationCount == 0, "Cannot set StreamListener value '" + methodAnnotatedInboundName + "' when using" +
+					" @Output annotation as method parameter. Use @Input method parameter annotation to specify inbound value instead");
+		}
+		else {
+			Assert.isTrue(inputAnnotationCount >= 1, "No input destination is configured. Use either a @StreamListener argument or @Input");
+		}
+		if (StringUtils.hasText(methodAnnotatedOutboundName)) {
+			Assert.isTrue(outputAnnotationCount == 0, "Cannot set both Output (@Output/@SendTo) method annotation value '" + methodAnnotatedOutboundName +
+					"' and @Output annotation as a method parameter");
+		}
+	}
+
+	private void assertStreamListenerMessageHandlerMethod(Method method, String inboundName) {
+		Assert.isTrue(StringUtils.hasText(inboundName), "No input destination is configured. Use @StreamListener argument");
+		int methodArgumentsLength = method.getParameterTypes().length;
+		if (methodArgumentsLength > 1) {
+			int numAnnotatedMethodParameters = 0;
+			int numPayloadAnnotations = 0;
+			for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+				MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+				if (methodParameter.hasParameterAnnotations()) {
+					numAnnotatedMethodParameters++;
+				}
+				if (methodParameter.hasParameterAnnotation(Payload.class)) {
+					numPayloadAnnotations++;
 				}
 			}
-			Assert.notNull(arguments[parameterIndex],
-					"Cannot convert argument " + parameterIndex + " of " + method + "from " + targetBean.getClass()
-							.toString() + " to " + parameterType.toString());
+			Assert.isTrue(methodArgumentsLength <= numAnnotatedMethodParameters && numPayloadAnnotations <= 1,
```

Comment:
Sorry, this was intended to be `methodArgumentsLength >= numAnnotatedMethodParameters`